design/design-generic-references.md

```diff
@@ -0,0 +1,319 @@
+    - apiVersion: example.com/v1
+      kind: Foo
+      name: foo
+      fieldPath: spec.port
+      id: port # optional variable name in template context
+  mapping:
+    value: "https://{{ .host }}:{{ .port }}"
+```
+
+The used templating engine in `mapping.value` is [Go templates]. Each object 
+reference with assigned `id` becomes available under that name in the template context.
+A number of useful transformation functions could be made available to that context.
+
+### Advantages
+
+* All referred values are declared on managed resources using already familiar
+  mechanism, keeping their discovery simple
+* Value extraction is detached from its consumption. The same `Referable` can be
+  used by multiple managed resources
+* `Referable` instances can be watched and used by non-Crossplane controllers as
+  well
+* `Referable` instances can be properly garbage collected if owner
+  or [`Usage`](https://github.com/crossplane/crossplane/blob/master/design/one-pager-generic-usage-type.md)
+  instances are declared
+* Multiple objects/fields can be referred as value sources
+* Value can be transformed using a Go template
+
+### Disadvantages
+
+* Providers might need to get read access to `Referable` type 
+  in order to resolve generic cross-resource references
+
+### Implementation
+
+It consists of two parts:
+
+* Adding new Crossplane controller for the `Referable` type
+* Update [crossplane-runtime](https://github.com/crossplane/crossplane-runtime/)
+  to support usage of `Referable` instances
+
+#### Referable Controller
+
+The controller is responsible for:
+
+* Defining proper `Usage` instance to guard against improper deletion
+  of `Referable` instances
+* Retrieving the declared objects and the field value, based on the provided
+  references
+* Transforming the value (if requested) and exposing it under `status.value`
+* Marking the instance as ready
+
+We assume currently that the value does not change after it gets exposed.
+Allowing updates and propagating them to the managed resources is out
+of the scope for this proposal version.
+
+No additional RBAC rules are needed, since Crossplane already poses very broad
+permissions within the cluster.
+
+#### Crossplane-runtime Changes
+
+* [APIResolver](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reference/reference.go#L280)
+  should be enriched to support `Referable` instances
+* If additional references need to be declared via the `spec.refs` block or
+  annotations, the [ResolveReferences function generator](https://github.com/crossplane/crossplane-tools/blob/master/internal/method/resolver.go#L33)
+  should be enriched to support them
+* After fetching values, proper `Usage` instance need to be defined to guard
+  against improper deletion
+
+Finally, providers need to be upgraded to use the new version of crossplane-runtime 
+and for each type, the `ResolveReference` function needs to be regenerated.
+
+## Alternatives Considered
```

Comment:
- this proposal seems to make the same mistakes that the composition patch and transformation (P&T) has done, and that eventually led to the invention of functions:
  - desirable transformations are open ended and are never enough
  - patching by field paths is a very low-level mechanism, operating on low-level data-type rather than being MR kind aware
  - references don't solve the order problem: (a) no way to order without the MR author having prepared for a reference field (b) dependencies might be more than just a field in an object (think of probes).
- in general, this proposal seems to clone the P&T idea into the object system. Do we really want that? Then the new approach competes with compositions. This is kind of the goal stated in the beginning of the doc. I miss a discussion why this is an acceptable goal, i.e. a motivation section.
- the core motivation behind this proposal seems to be that compositions are too slow (e.g. not watch-driven, but run in 60s interval). Maybe we should fix the latter.

Reply:
+1
This feels like it's adding one more layer of indirection to Compositions and feels overlapping with both the `Usage` resource introduced by https://github.com/crossplane/crossplane/pull/4215 and a little bit with observe only resources too.

Reply:
It would be good to define and agree on the problem before talking about the solution.

Let's take the `VPC` / `Subnet` example. If someone wants to create a `Subnet` in a `VPC` created/managed by XP, they need a mechanism to read the `id` from `VPC` and put it into the `Subnet` spec. Today, this is possible in one of the following two ways:

**a.** Using cross-resource references where they can either use MR name for `VPC` (e.g. `vpcIdRef.name`) or labels/controllerRef (e.g. `vpcIdRef.selector`).
**b.** Using composition patches, but with the following limitations:
  - `VPC` and `Subnet` must exist in the same composition.
  - `id` needs to be patched to XR first, (e.g. `status.vpcId`) and then patched to the `Subnet` (toFieldPath: `spec.forProvider.vpcId`)

Typically, in compositions we see that the option (a) is used with `matchControllerRef` where the `id` of the `VPC` with the same controller ref (owned by the same XR) is extracted and used in `Subnet`:

```
            vpcIdSelector:
              matchControllerRef: true
```

Then, what is wrong? There are two problems invalidating both of the available options as such:

**option (a):** Not every MR field directly maps to a single type, so existing XRR falls short. There are a lot of issue references in the [original ticket](https://github.com/crossplane/crossplane/issues/1770), but one example is the `destinationArn` on a `SubscriptionFilter` resource where destination could be one of the [multiple types](https://github.com/upbound/upjet/issues/95#issuecomment-1487739037).
**option (b):** Resources may not be part of the same composite.

So, we need a solution here. If someone creates a `SubscriptionFilter` and a `DeliveryStream` (using the [same example](https://github.com/upbound/upjet/issues/95#issuecomment-1487739037)) with Crossplane but not necessarily within the same composition, they need a way to fill `spec.forProvider.destinationArn` on `SubscriptionFilter` with the arn of the `DeliveryStream` they intend to. 

> this proposal seems to make the same mistakes that the composition patch and transformation (P&T) has done, and that eventually led to the invention of functions:

I agree that the problem we're trying to solve here is **similar** to what (P&T) has done, but there is that fundamental difference that makes it impossible to solve with compositions or functions, which is `resources may not be part of the same composite`. I can see two ways to overcome this, both with some implications:

- If you want to patch from a resource that is not part of the same composition, define an `ObserveOnly` resource in the same composition and patch from it.
- Functions could access to k8s API and read anything it wants without being limited to `observed` state.

> desirable transformations are open ended and are never enough

Yes. While we lack very basics today, like _take that ARN and put it here_, it would eventually need more and more, just like we see in composition patches. Leveraging and relying on some [external](https://jmespath.org/),[tool](https://github.com/crossplane/crossplane/issues/4351) might help here.

> references don't solve the order problem: (a) no way to order without the MR author having prepared for a reference field (b) dependencies might be more than just a field in an object (think of probes).

I am not aware of a real (creation) ordering problem today where we are good enough with eventual consistency.
We had the deletion ordering problem and tackling it with the `Usage` type. 
I am not claiming it will not complicate anything but we can still stay on the "declarative" world by saying, _this value needs to be constructed this way and eventually patched there._

> in general, this proposal seems to clone the P&T idea into the object system. Do we really want that?

Not sure if the solution should be the one proposed here, but we need a solution to the problem defined above.

> the core motivation behind this proposal seems to be that compositions are too slow (e.g. not watch-driven, but run in 60s interval). Maybe we should fix the latter.

Not sure how you came up with this assumption, but I am not aware of such a motivation for Generic XRR. IIRC, compositions not being watch-driven came up recently in the composition function (re)design; however, Generic XRR has been an [open issue](https://github.com/crossplane/crossplane/issues/1770) for more than 2 years. One can claim that, if we could live without it for 2 years, then maybe it is not that an important problem to solve :) I would love to hear from the community on that, and I believe it is not a blocker mostly because there are workaround/hacks there which could unblock use cases but ultimately complicate things further.

Reply:
I'd like to add a slightly different perspective based on my experience using `EnvironmentConfigs`.  I think there is a fair amount of overlap in the usage scenarios, and for me personally the `EnvironemtConfig` model is simpler and easier to understand and use than any of the other proposals for separate resource references that I have seen.

One of the common use cases we have found for `EnvironmentConfig` is making information available from one Composite to another.  If Composite A creates a set of resources and wants to make some information regarding those resources available to any other Composite/Composition that may need them, it can create an `EnvironmentConfig` object with the appropriate labels and the information to be "published".  Other Composites can retrieve these `EnvironmentConfigs` as needed in their Compositions using the appropriate label selectors.

The process of finding the appropriate `EnvironmentConfigs` and loading them into the Composition `environment` creates what is essentially a runtime context for that Composite.  The user (composition author) can leverage this context to pass information between resources and can reference external data using `EnvironmentConfigs`.  The structure of the local context is defined by the author and information is placed in specific locations by name.  This model works extremely well and provides necessary functionality.

So why limit the `environment` external data references to EnvironmentConfigs?  Why not allow the `environment` specification to include references to any existing Kubernetes object, which can be retrieved by the same mechanism that resolves EnvironmentConfigs today?  If an author wants to retrieve data from an existing XR or MR or ConfigMap or Secret they could specify the G/V/K information required to find the object along with a name or label selector,  and a path within the object to the data they want, and that information can be populated in the `environment` in the specified location.

From the author's perspective this would let me put all of my external data references in one place, and they can be retrieved dynamically at runtime as needed if they exist, or they can eventually reconcile as they become available.

I think one of the overarching goals needs to be keeping the user experience as simple as possible, and requiring a specific resource definition for every reference feels like it is adding complexity for the user and making the composition more difficult to develop and maintain.

Reply:
Per https://github.com/crossplane/crossplane/issues/1770#issuecomment-1627868262 I share the concern that generic references grows to be something like P&T Composition.

Also, per https://github.com/crossplane/crossplane/issues/1770#issuecomment-1633270957 I'm pretty confident that today's hardcoded style of references isn't going to scale. It's too manual to support, too inflexible for users, and creates a coupling between providers when one provider references a type defined by another.

I could be convinced that the answer is to just not have MR-level references at all if we can find a clean way to address this at the Composition level. I would expect such an approach to be controversial though. I suspect there's a not-insignificant amount of folks who don't use Composition today who would be forced to if Composition was the only way to derive one

Reply:
> I could be convinced that the answer is to just not have MR-level references at all if we can find a clean way to address this at the Composition level. I would expect such an approach to be controversial though. I suspect there's a not-insignificant amount of folks who don't use Composition today who would be forced to if Composition was the only way to derive one

This proposal talks about that path as well. No matter what strategy/convention we would like to apply to MR-level reference, at the end of the day, it needs to be implemented by providers. On the other hand, solving the problem via compositions requires no changes at provider side.